arch/rv32i/src/pmp.rs

```diff
@@ -490,210 +683,472 @@ impl<const MAX_AVAILABLE_REGIONS_OVER_TWO: usize> kernel::platform::mpu::MPU
+        mpu.remove_memory_region(region_0, &mut config)
+            .expect("Failed to remove valid MPU region allocation");
+
+        for region @ (memory_start, memory_size, length, perms) in overlap_region_0_tests.iter() {
+            let allocation_res =
+                mpu.allocate_region(*memory_start, *memory_size, *length, *perms, &mut config);
+
+            match allocation_res {
+                Some(region) => {
+                    mpu.remove_memory_region(region, &mut config)
+                        .expect("Failed to remove valid MPU region allocation");
+                }
+                None => {
+                    panic!(
+                        "Failed to allocate region that does not overlap and \
+                         should meet alignment constraints: {:?}",
+                        region
+                    );
                 }
             }
         }
 
-        let region_num = config.unused_kernel_region_number(self.locked_region_mask.get())?;
+        // Make sure we can technically allocate a memory region that overlaps
+        // with the kernel part of the `app_memory_region`.
+        //
+        // It is unclear whether this should be supported.
+        let region_2 = mpu
+            .allocate_region(
+                0xd0000000 as *const u8,
+                0x10000000,
+                0x10000000,
+                Permissions::ReadWriteOnly,
+                &mut config,
+            )
+            .unwrap();
+        assert!(region_2.start_address() == 0xd0000000 as *const u8);
+        assert!(region_2.size() == 0x10000000);
+
+        // Now, we can grow the app memory break into this region:
+        mpu.update_app_memory_region(
+            0xd0000004 as *const u8,
+            0xd8000000 as *const u8,
+            Permissions::ReadWriteOnly,
+            &mut config,
+        )
+        .expect("Failed to grow the app memory region into an existing other MPU region");
+
+        // Now, we have two overlapping MPU regions. Remove `region_2`, and try
+        // to reallocate it as `region_3`. This should fail now, demonstrating
+        // that we managed to reach an invalid intermediate state:
+        mpu.remove_memory_region(region_2, &mut config)
+            .expect("Failed to remove valid MPU region allocation");
+        assert!(mpu
+            .allocate_region(
+                0xd0000000 as *const u8,
+                0x10000000,
+                0x10000000,
+                Permissions::ReadWriteOnly,
+                &mut config,
+            )
+            .is_none());
+    }
 
-        // Logical region
-        let mut start = memory_start as usize;
-        let mut size = memory_size;
+    #[test]
+    pub fn test_app_memory_alloc() {
+        // TODO! implement more test cases, such as:
+        //
+        // - Try to update the app memory break with an invalid pointer below
+        //   its allocation's start address.
```

Comment:
TODO: More test cases here! Especially w.r.t. to alignment & avoiding memory region overrun.